internal/util/merge/merge3.go

```diff
@@ -268,3 +272,74 @@ func metadataComment(node *yaml.RNode) string {
 	}
 	return mf.Key.YNode().LineComment
 }
+
+// resourceHandler is an implementation of the ResourceHandler interface from
+// kyaml. It is used to decide how a resource should be handled during the
+// 3-way merge. This differs from the default implementation in that if a
+// resource is deleted from upstream, it will only be deleted from local if
+// there is no diff between origin and local.
+type resourceHandler struct {
+	keptResources []*yaml.RNode
+}
+
+func (r *resourceHandler) Handle(origin, upstream, local *yaml.RNode) (filters.ResourceMergeStrategy, error) {
+	var strategy filters.ResourceMergeStrategy
+	switch {
+	case origin == nil && upstream == nil && local != nil:
+		strategy = filters.KeepDest
+	case upstream != nil && local == nil:
+		strategy = filters.KeepUpdated
+	case upstream == nil && local == nil:
+		strategy = filters.Skip
+	case origin != nil && upstream == nil:
+		equal, err := r.equals(origin, local)
+		if err != nil {
+			return strategy, err
+		}
+		if equal {
+			strategy = filters.Skip
+		} else {
+			r.keptResources = append(r.keptResources, local)
+			strategy = filters.KeepDest
+		}
+	case origin != nil && local == nil:
+		strategy = filters.Skip
+	default:
+		strategy = filters.Merge
+	}
+	return strategy, nil
+}
+
+func (*resourceHandler) equals(r1, r2 *yaml.RNode) (bool, error) {
+	// We need to create new copies of the resources since we need to
+	// mutate them before comparing them.
+	r1Clone, err := yaml.Parse(r1.MustString())
+	if err != nil {
+		return false, err
+	}
+	r2Clone, err := yaml.Parse(r2.MustString())
+	if err != nil {
+		return false, err
+	}
+
+	// The resources include annotations with information used during the merge
+	// process. We need to remove those before comparing the resources.
+	if err := stripKyamlAnnos(r1Clone); err != nil {
+		return false, err
+	}
+	if err := stripKyamlAnnos(r2Clone); err != nil {
+		return false, err
+	}
+
+	return r1Clone.MustString() == r2Clone.MustString(), nil
+}
+
+func stripKyamlAnnos(n *yaml.RNode) error {
+	for _, a := range []string{mergeSourceAnnotation, kioutil.PathAnnotation, kioutil.IndexAnnotation} {
```

Comment:
There is id annotation, not sure if it relevant here.

Reply:
Which annotation are you referring to? In this case we don't want to remove all annotations, only the ones that has been added temporarily as part of the merge process.